File: 0_数组/改变数组/665. 非递减数列.js
// 给你一个长度为 n 的整数数组，请你判断在 最多 改变 1 个元素的情况下，该数组能否变成一个非递减数列。
/**
 * @param {number[]} nums
 * @return {boolean}
 * @link https://leetcode-cn.com/problems/non-decreasing-array/solution/3-zhang-dong-tu-bang-zhu-ni-li-jie-zhe-d-06gi/
 * @summary
 * 当 nums[i] 破坏了数组的单调递增时，即 nums[i] < nums[i - 1]  时，为了让数组有序，我们发现一个规律
 * 1. 当 i = 1 ，那么修改 num[i- 1] ，不要动 nums[i]
 * 2.当 i > 1 时，我们应该优先考虑把 nums[i - 1] 调小到 >= nums[i - 2] 并且 <= nums[i]。
 * 3. 当 i > 1 且 nums[i] < nums[i - 2] 时，我们无法调整 nums[i - 1] ，我们只能调整 nums[i] 到 nums[i - 1] 。
 * 即：尽量调前面，实在不行调后面
 */
var checkPossibility = function (nums) {
  let res = 0
  const n = nums.length
  for (let i = 1; i < n; i++) {
    if (nums[i] < nums[i - 1]) {
      res++
      if (i === 1 || nums[i] >= nums[i - 2]) {
        nums[i - 1] = nums[i]
      } else {
        nums[i] = nums[i - 1]
      }
    }
  }

  return res <= 1
}

File: leetcode/665. 非递减数列.py
# 665. 非递减数列
# 给你一个长度为 n 的整数数组 nums ，请你判断在 最多 改变 1 个元素的情况下，该数组能否变成一个非递减数列。

from typing import List


class Solution:
    def checkPossibility(self, nums: List[int]) -> bool:
        """
        贪心扫描，记录逆序对次数 count。遇到 nums[i] < nums[i-1] 时：
          - 如果已经有一次修改，直接返回 False。
          - 否则尝试“修正”一次：
            • 如果 i<2 或 nums[i] >= nums[i-2]，则把 nums[i-1] 调低到 nums[i]（等价于修改前一个元素）。
            • 否则把 nums[i] 调高到 nums[i-1]（修改当前元素）。
        最终若 count ≤ 1 则可行。
        时间 O(n)，空间 O(1)。
        """
        count = 0
        for i in range(1, len(nums)):
            if nums[i] < nums[i - 1]:
                count += 1
                if count > 1:
                    return False
                # 尝试修改 nums[i-1] 或 nums[i]
                if i < 2 or nums[i] >= nums[i - 2]:
                    # 安全地把前一个元素调低
                    nums[i - 1] = nums[i]
                else:
                    # 否则只能把当前元素调高
                    nums[i] = nums[i - 1]
        return True
